src/libPMacc/include/cuSTL/algorithm/kernel/run-time/Foreach.hpp

```diff
@@ -52,6 +53,28 @@ namespace kernel
 namespace RT
 {
 
+/** Return a suitable cuda blockDim for a given gridDim.
+ *
+ * @param gridDim 1D, 2D or 3D grid size
+ * @return cuda blockDim
+ */
+template<int dim>
+math::Size_t<3> getBestCudaBlockDim(math::Size_t<dim> gridDim)
+{
+    math::Size_t<3> result = math::Size_t<3>::create(1);
+
+    /* The greatest common divisor of each component of the volume size
+     * and a certain power of two value gives the best suitable block size */
+    boost::math::gcd_evaluator<size_t> gcd; /* greatest common divisor */
+    size_t maxValues[] = {16, 16, 4}; /* maximum values for each axis */
```

Comment:
> it's a very rarely used function

the first point is not an argument that I can accept if I use cuSTL :)

> tuning maxValues is just an optimization

the second  one is not only an optimization: a new integrated SoC might for example not support so many threads or a port to alpaka would also like to limit the max threads to fit on a multi-core CPU.

the maximum number of threads is a global device attribute and should be adjustable centrally in either cuSTL or libPMacc.